File: game/player.js
// const AVAILABLE_COLORS = ['red', 'blue', 'purple', 'green', 'yellow', 'orange', 'X'];

function Player(player_id, player_name, player_location, player_color, is_mr_x) {
    const id = player_id; // websocket connection ID
    const name = player_name;
    let color = player_color;
    let isMrX = is_mr_x;
    let location = player_location;

    let tickets = isMrX ? {
        taxi: 4,
        bus: 3,
        underground: 3,
        black: 5,
        times_two: 2
    } : {
        taxi: 10,
        bus: 8,
        underground: 4
    };

    // getters
    this.getID = () => id;
    this.getName = () => name;
    this.getColor = () => color;
    this.isMrX = () => isMrX;
    this.getLocation = () => location;
    this.getTickets = (type) => {
        switch (type) {
            case 'taxi':
                return tickets.taxi;
                break;
            case 'bus':
                return tickets.bus;
                break;
            case 'underground':
                return tickets.underground;
                break;
            case 'black':
                if (isMrX) return tickets.black;
                break;
        }
    }

    //setters
    this.setColor = c => {
        color = c;
    };
    this.setMrX = () => {
        isMrX = true;
        color = 'X';
    };
    this.unsetMrX = (c) => {
        isMrX = false;
        color = c;
    };
    this.setLocation = loc => location = 1 <= loc && loc <= 200 ? loc : location;
    this.discard = (type) => {
        switch (type) {
            case 'taxi':
                tickets.taxi = tickets.taxi === 0 ? tickets.taxi : tickets.taxi - 1;
                break;
            case 'bus':
                tickets.bus = tickets.bus === 0 ? tickets.bus : tickets.bus - 1;
                break;
            case 'underground':
                tickets.underground = tickets.underground === 0 ? tickets.underground : tickets.underground - 1;
                break;
            case 'black':
                if (isMrX) tickets.black = tickets.black === 0 ? tickets.black : tickets.black - 1;
                break;
            case 'times_two':
                if (isMrX) tickets.times_two = tickets.times_two === 0 ? tickets.times_two : tickets.times_two - 1;
                break;
        }
    }

    this.gain = (type) => { // Gain tickets discarded by detectives
        if (isMrX) {
            switch (type) {
                case 'taxi':
                    tickets.taxi++;
                    break;
                case 'bus':
                    tickets.bus++;
                    break;
                case 'underground':
                    tickets.underground++;
                    break;
            }
        }
    }
}

module.exports = Player;

File: scotlandyard/scotlandyardgame/engine/player.py
class Player:

    def __init__(self, player_id: str, player_name: str, player_location: int, player_color: str, is_mr_x: bool) -> None:
        # private
        self.__player_id: str = player_id
        self.__player_location: int = player_location
        self.__is_mr_x: bool = is_mr_x
        self.__tickets: dict[str, int] = {
            "taxi": 4,
            "bus": 3,
            "underground": 3,
            "black": 5,
            "times_two": 2
        } if is_mr_x else {
            "taxi": 10,
            "bus": 8,
            "underground": 4
        }

        # public
        self.name: str = player_name
        self.color: str = player_color

    # getters

    @property
    def ID(self) -> str:
        return self.__player_id

    @property
    def location(self) -> int:
        return self.__player_location

    @property
    def is_mr_x(self) -> bool:
        return self.__is_mr_x

    # setters

    @ID.setter
    def ID(self, newID: str):
        raise AttributeError("ID assignment not allowed.")

    @location.setter
    def location(self, newLocation: int):
        self.__player_location = newLocation if 1 <= newLocation <= 200 else self.__player_location

    @is_mr_x.setter
    def is_mr_x(self, value: bool, color: str = 'X'):
        if not value and color == 'X':
            raise ValueError("cannot assign color X to non-Mr. X Player.")
        self.__is_mr_x = value
        self.color = 'X' if value else color

    # methods

    def getTickets(self, type: str) -> int:
        """returns number of tickets of type ```type``` available to this player."""
        return self.__tickets[type]

    def discard(self, type):
        """player uses a ticket."""
        self.__tickets[type] -= 1 if self.__tickets[type] else 0

    def gain(self, type):
        if not self.is_mr_x:
            raise TypeError("non-Mr. X Player cannot gain a ticket.")
        if type not in ["taxi", "bus", "underground"]:
            raise ValueError(f"cannot gain ticket of type '{type}.'")
        self.__tickets[type] -= 1
